src/cargo/sources/registry/mod.rs

```diff
@@ -740,15 +750,11 @@ impl<'cfg> Source for RegistrySource<'cfg> {
             let mut called = false;
             ready!(self.index.query_inner(
                 dep.package_name(),
-                dep.version_req(),
+                &req,
                 &mut *self.ops,
                 &self.yanked_whitelist,
                 &mut |s| {
-                    let matched = match kind {
-                        QueryKind::Exact => dep.matches(&s),
-                        QueryKind::Fuzzy => true,
-                    };
-                    if matched {
+                    if dep.matches(&s) || kind == QueryKind::Fuzzy {
```

Comment:
The downside to this refactor (separate commit?) is that we always evaluate `dep.matches(&s)` even if we don't need to.

However, unsure how much that matters as I'm assuming `Exact` is the more common case.

The other downside is if we add new cases to `kind`, the compiler won't yell at us to update this code (which is why I used explicit matches when I did that refactor)

Reply:
Backed out.